Design/SANS/BeamCentreFinderDesign.md

```diff
@@ -0,0 +1,271 @@
+
+The different approaches of ISIS and ANSTO are currently not compatible. They can be summarized as:
+
+* ISIS makes use of a so-called ReductionSingleton which has a series of registered reduction steps. During the reduction, the ReductionSingleton() carries state (mainly via the custom ISIS instrument classes). Intermediate results are stored in the Analysis Data Service (ADS).
+
+* ANSTO's reduction implementation is currently still work in progress, but will eventually make use of so-called Workflow-Algorithms which do not populate the ADS or carry state information. This will be a cleaner, more maintainable implementation which is easier to test and verify. It is desirable that in the long run ISIS SANS switches to such an implementation.
+
+This document outlines the required changes to make the *BCF* useable in a more general, faciltity- and instrument-agnostic way.
+
+The benefits of introducing a more general BCF are:
+
+* Avoiding a facility-based implementation and having this functionality in a dedicated location will make the BCF much more maintainable and it will be easier to introduce changes which affect all facilities.
+* A *BCF* which does not depend on facility- and instrument-specifics will make this functionality reusable for new instruments and facilities which might join Mantid in the future.
+* Removing the dependence of the ReductionSingleton will allow for easier unit testing of the BCF.We currently have one system test and zero unit tests for the *BCF* functionality. 
+
+
+The static details of the current implementation (ie a class diagram) can be seen here:
+
+![alt text](beam_centre_finder_classes.png)
+
+
+The workflow of the current implementation is illustrated below. It is separated into two parts, one which is repsonsible for stepping through the parameter space
+![alt text](FindBeamCentre_py.png)
+
+and one which illustrates how the residuals -- a measure of our search quality -- are calculated. This is achieved by the *SeekCentre* operation:
+
+![alt text](centre_finder_py.png)
+
+### Proposed solution
+In general terms we want to hide any solution-specific details in the *BCF*. Currently there is a layer of abstration which hides the instrument-specific details, e.g. LARMOR uses an instrument rotation instead of a translation. Extending this abstraction further to make the functionality also facility-agnostic will allow us to apply the *BCF* to any facility or instrument.
+
+One way of achieving this is a generic interface which is discussed below. In this case each instrument\facility needs to provide an implementation for the generic interface. This implementation is then injected into the BCF. 
+
+Another, similar way, would be to use a Template Method approach. We would have to create a class-based version of the BCF. Each instrument/facility then implements the methods which need to be overwritten.
```

Comment:
Worth mentioning this Template Method option. But I might mention something about why Strategy injection would be superior.